model-optimizer/extensions/ops/scatternd.py

```diff
@@ -0,0 +1,101 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+
+from mo.front.common.partial_infer.utils import int64_array
+from mo.graph.graph import Node, Graph
+from mo.ops.op import Op
+
+
+class ScatterNDBase(Op):
+    enabled = False
+
+    op = op_type = None
+    version = None
+
+    def __init__(self, graph: Graph, attrs: dict):
+        assert self.op is not None and self.op_type is not None and self.version is not None, \
+            'Please use specialized ScatterNDBase operation class, ScatterNDBase is base class'
+
+        mandatory_props = {
+            'op': self.op,
+            'type': self.op_type,
+            'version': self.version,
+
+            'infer': self.infer,
+
+            'in_ports_count': 3,
+            'out_ports_count': 1,
+        }
+        super().__init__(graph, mandatory_props, attrs)
+
+    @staticmethod
+    def infer(node: Node):
+        node_name = node.soft_get('name', node.id)
+
+        input_shape = node.in_port(0).data.get_shape()
+        indices_shape = node.in_port(1).data.get_shape()
+        updates_shape = node.in_port(2).data.get_shape()
+        assert input_shape is not None and updates_shape is not None and indices_shape is not None, \
+            'The node "{}" input shape is None'.format(node_name)
+
+        # check that shapes are correct
+        # 1. ranks of both input and indices must be at least 1
+        assert len(input_shape) >= 1 and len(indices_shape) >= 1, \
+            'The node "{}" input and indices ranks must be at least 1'.format(node_name)
+           
+        # 2. the last dimension of indices shape must be at most a rank of input
+        assert indices_shape[-1] <= len(input_shape), \
+            'The last dimension of indices shape must be at most a rank of input for the node "{}"'.format(node_name)
+
+        # 3. updates is a tensor of shape indices_shape[:-1] + input_shape[indices_shape[-1]:]
+        expected_updates_shape = np.concatenate((indices_shape[:-1], input_shape[indices_shape[-1]:]), axis=0)
+        if np.array_equal(expected_updates_shape, int64_array([])):
+            expected_updates_shape = 0 # updates is a scalar
+        assert np.array_equal(updates_shape, expected_updates_shape), \
+            'The updates shape must be equal to indices_shape[:-1] + input_shape[indices_shape[-1]:] for the node "{}"'.format(node_name)
+
+        node.out_port(0).data.set_shape(input_shape)
+
+    @staticmethod
+    def type_infer(node: Node):
+        assert node.in_port(0).get_source().get_data_type() == node.in_port(2).get_source().get_data_type(), \
+            'The data type of the first and the third inputs must be equal for the node {}'.format(node.name)
+        node.out_port(0).set_data_type(node.in_port(0).get_source().get_data_type())
```

Comment:
Should it be `node.out_port(0).set_data_type(node.in_port(0).get_data_type())`? Without the `get_source()`, because the node producing input to the `node.in_port(0)` may have multiple outputs.

Reply:
Added one more test with a scalar. Now we have for value infer as well.